pinot-core/src/main/java/com/linkedin/pinot/core/minion/BackfillDateTimeColumn.java

```diff
@@ -0,0 +1,258 @@
+   * It simply reads the records from the base record reader, and adds a new field according to the
+   * dateTimeFieldSpec
+   */
+  public class BackfillDateTimeRecordReader extends BaseRecordReader {
+
+    private RecordReader _baseRecordReader;
+    private TimeFieldSpec _timeFieldSpec;
+    private DateTimeFieldSpec _dateTimeFieldSpec;
+    private int _numRows = Integer.MAX_VALUE;
+    private int _nextRow;
+
+    /**
+     * Constructor to read from the base record reader, but limit the number of rows read
+     * This is useful when backfilling star tree segments, so that we read only total raw docs
+     * @param baseRecordReader
+     * @param timeFieldSpec
+     * @param dateTimeFieldSpec
+     * @param maxNumRows
+     * @throws IOException
+     * @throws ConfigurationException
+     */
+    public BackfillDateTimeRecordReader(RecordReader baseRecordReader, TimeFieldSpec timeFieldSpec,
+        DateTimeFieldSpec dateTimeFieldSpec, int maxNumRows) {
+      this(baseRecordReader, timeFieldSpec, dateTimeFieldSpec);
+      _numRows = maxNumRows;
+    }
+
+    public BackfillDateTimeRecordReader(RecordReader baseRecordReader, TimeFieldSpec timeFieldSpec,
+        DateTimeFieldSpec dateTimeFieldSpec) {
+      _baseRecordReader = baseRecordReader;
+      _timeFieldSpec = timeFieldSpec;
+      _dateTimeFieldSpec = dateTimeFieldSpec;
+      _nextRow = 0;
+    }
+
+    @Override
+    public void init() throws Exception {
+      _baseRecordReader.init();
+      _nextRow = 0;
+    }
+
+    @Override
+    public void rewind() throws Exception {
+      _baseRecordReader.rewind();
+      _nextRow = 0;
+    }
+
+    @Override
+    public boolean hasNext() {
+      return _nextRow < _numRows && _baseRecordReader.hasNext();
+    }
+
+    /**
+     * Reads the schema from the baseRecordReader and adds/updates dateTimeFieldSpec to it, if not
+     * already present {@inheritDoc}
+     * @see com.linkedin.pinot.core.data.readers.RecordReader#getSchema()
+     */
+    @Override
+    public Schema getSchema() {
+      Schema schema = _baseRecordReader.getSchema();
+      // base record reader doesn't have _dateTimeFieldSpec
+      if (schema.getDateTimeNames() == null
+          || !schema.getDateTimeNames().contains(_dateTimeFieldSpec.getName())) {
+        schema.addField(_dateTimeFieldSpec);
+      }
+      // base record reader has dateTimeFieldSpec, but with different configs
+      // create schema with updated dateTimeFieldSpec
+      else if (!_dateTimeFieldSpec.equals(schema.getDateTimeSpec(_dateTimeFieldSpec.getName()))) {
+        Schema newSchema = new Schema();
+        for (DimensionFieldSpec dimensionFieldSpec : schema.getDimensionFieldSpecs()) {
+          newSchema.addField(dimensionFieldSpec);
+        }
+        for (MetricFieldSpec metricFieldSpec : schema.getMetricFieldSpecs()) {
+          newSchema.addField(metricFieldSpec);
+        }
+        newSchema.addField(schema.getTimeFieldSpec());
```

Comment:
Dont u have to re-add the existing DateTimeFieldSpec (that do not match the name of the new datetime field), if any?

Reply:
Yes looks like I made an incorrect assumption that there will be only 1. Fixed this.